db/migrate/20171116110243_migrate_existing_scenarios_for_public_transport_changes.rb

```diff
@@ -0,0 +1,98 @@
+    assumptions = {
+      'transport_passenger_train_using_coal_share' => 'transport_freight_train_using_coal_share',
+      'transport_passenger_train_using_diesel_mix_share' => 'transport_freight_train_using_diesel_mix_share',
+      'transport_passenger_train_using_electricity_share' => 'transport_freight_train_using_electricity_share',
+      'transport_truck_using_compressed_natural_gas_share' => 'transport_bus_using_compressed_natural_gas_share',
+      'transport_truck_using_diesel_mix_share' => 'transport_bus_using_diesel_mix_share',
+      'transport_truck_using_electricity_share' => 'transport_bus_using_electricity_share',
+      'transport_truck_using_gasoline_mix_share' => 'transport_bus_using_gasoline_mix_share',
+      'transport_truck_using_hydrogen_share' => 'transport_bus_using_hydrogen_share',
+      'transport_truck_using_lpg_share' => 'transport_bus_using_lpg_share',
+      'transport_car_using_electricity_share' => 'transport_motorcycle_using_electricity_share'
+    }
+
+    corrections = {
+      'transport_motorcycle_using_electricity_share' => 'transport_motorcycle_using_gasoline_mix_share'
+    }
+
+    # Pruned input keys
+    removed = [
+      'transport_useful_demand_ship_kms',
+      'transport_useful_demand_trains',
+      'transport_useful_demand_planes',
+      'transport_truck_using_compressed_natural_gas_efficiency'
+    ]
+
+    scenarios = Scenario.where("`user_values` IS NOT NULL AND `balanced_values` IS NOT NULL AND `user_values` != '--- !ruby/hash:ActiveSupport::HashWithIndifferentAccess {}\n' AND `balanced_values` != '--- !ruby/hash:ActiveSupport::HashWithIndifferentAccess {}\n'")
+
+    bar = ProgressBar.new(scenarios.count)
+
+    scenarios.each do |scenario|
+      %i(user_values balanced_values).each do |inputs_attribute|
+        # Step 1: Translations
+        migrated = scenario.public_send(inputs_attribute).each_with_object({}) do |(key, val), obj|
+          obj[(renamed[key.to_s] || key.to_s).to_sym] = val
+        end
+
+        # Step 2: include assumptions
+        migrated = migrated.each_with_object({}) do |(key, val), obj|
+          if assumed_key = assumptions[key.to_s]
+            obj[assumed_key.to_sym] = val
+          end
+
+          obj[key] = val
+        end
+
+        # Step 3: Corrections for motorcycles
+        migrated = migrated.each_with_object({}) do |(key, val), obj|
+          if fix = corrections[key.to_s]
+            obj[fix] = (100.0 - val)
+          end
+
+          obj[key] = val
+        end
+
+        # Step 4: Remove old one's
+        migrated = migrated.except(*removed.map(&:to_sym))
+
+        # Actual migration
+        scenario.public_send("#{ inputs_attribute }=", migrated)
+      end
+
+      if scenario.valid?
+        scenario.save
+        bar.increment!
+      else
+        a = "#{ DateTime.now } - #{ scenario.id } can't be migrated because '#{ scenario.errors.messages }'\n"
```

Comment:
Are there any scenarios that cannot be migrated on the staging or production DBs?

I tend to think this is a fatal error and the migration should exit, but if the failing validation is unrelated to the new/renamed inputs you can just run `scenario.save(validate: false)` safe in the knowledge that the migration is working as expected. If the record was *already* invalid, I expect it is better for it to be invalid with the correct inputs than to be invalid with the *old* inputs.

Reply:
All the errors are related to non-existing `area_code`.s